lib/RWMutex.ts

```diff
@@ -64,38 +58,46 @@ export default class RWMutex {
    * @return {Promise} - Promise that resolves when the lock is acquired, rejects if an error occurs
    */
   async lock() {
-    // provides readable error messages, no need to catch and rethrow
-    const lock = await this._findOrCreateLock();
-
-    // if this clientID already has the lock, we re-enter the lock and return
-    if (lock.writer === this._clientID) {
-      return;
-    }
-
     // Loop and do the following:
     // 1. attempt to acquire the lock (must have no readers and no writer)
     // 2. if not acquired, sleep and retry
     let acquired = false;
     while (!acquired) {
       try {
+        // If no such lock exists, this will create it
+        // If a lock exists with this lockID with no readers and no writer, this will update it
+        // If a lock exists with this lockID with clientID as the writer and no readers,
+        // this will do nothing
+        // If a lock exists with this lockID with a different clientID as the writer or readers,
+        // this will throw an error which will be caught.  We will then retry.
         const result = await this._coll.updateOne(
           {
             lockID: this._lockID,
             readers: [],
-            writer: "",
+            $or: [
+              {
+                writer: "",
+              },
+              {
+                writer: this._clientID,
+              },
+            ],
           },
           {
             $set: {
               writer: this._clientID,
             },
-          }
+          },
+          { upsert: true },
         );
-        if (result.matchedCount > 0) {
+        if (result.matchedCount > 0 || result.upsertedCount > 0) {
           acquired = true;
           return;
         }
       } catch (err) {
-        throw new Error(`error aquiring lock ${this._lockID}: ${err.message}`);
+        if (!(err instanceof MongoError) || err.code !== 11000) {
```

Comment:
i know ur just moving code around, so no worries if you dont know what code 11000 is. but if you do, could we name it as a const? like `const ErrCodeMongoBlah = 11000` just so there is some description for it?